packages/eas-cli/src/commands/update/index.ts

```diff
@@ -241,14 +242,39 @@ export default class UpdatePublish extends EasCommand {
       const assets = filterExportedPlatformsByFlag(collectedAssets, platformFlag);
       realizedPlatforms = Object.keys(assets) as PublishPlatform[];
 
-      const uploadResults = await uploadAssetsAsync(
-        graphqlClient,
-        assets,
-        projectId,
-        (totalAssets, missingAssets) => {
-          assetSpinner.text = `Uploading (${totalAssets - missingAssets}/${totalAssets})`;
-        }
-      );
+      let lastUploadedStorageKeys = new Set<string>();
+      let lastAssetUploadResults: {
+        asset: RawAsset & { storageKey: string };
+        finished: boolean;
+      }[] = [];
+      let timeAtWhichToTimeout = Date.now() + 60 * 1000; // sixty seconds from now
+
+      const uploadResults = await Promise.race([
```

Comment:
with this approach after one promise is finished the second is still running which seems dangerous, e.g. after we throw a timeout upload is still happening. I suspect after both timeout or upload finish command ends quite quickly, so it's unlikely to cause problems, but this still feels wrong.

I don't really have a solution but maybe a few things to consider:
- In a function that checks for timeout you could run process exit (not sure if it makes sense here), it would avoid the uncertainty of what happens after.
- You could use the return value of callback to uploadAssetsAsync to tell the function to stop.
- After uploadResults is resolved or rejected you could set some flag and in the loop that checks for timeout check that flag and exit early.
- I feel like the same behavior would be more idiomatic if it was implemented using setTimeout/clearTimeout, but my attempts lead to more complicated code, so it might not make sense.

Reply:
Nice catch. I added a cancellationtoken-like object which is a concept in C#.